concurrent_test.go

```diff
@@ -292,138 +236,159 @@ func runWorker(t *testing.T, w worker, errCh chan error) (historyRecords, error)
 	return rs, err
 }
 
-type worker interface {
-	name() string
-	run() (historyRecords, error)
-}
-
-type readWorker struct {
+type worker struct {
 	id int
 	db *btesting.DB
 
 	bucket []byte
 	keys   []string
 
-	readTime duration
+	conf concurrentConfig
 
 	errCh  chan error
 	stopCh chan struct{}
 
 	t *testing.T
 }
 
-func (r *readWorker) name() string {
-	return fmt.Sprintf("readWorker-%d", r.id)
+func (w *worker) name() string {
+	return fmt.Sprintf("worker-%d", w.id)
 }
 
-func (r *readWorker) run() (historyRecords, error) {
+func (w *worker) run() (historyRecords, error) {
 	var rs historyRecords
 	for {
 		select {
-		case <-r.stopCh:
-			r.t.Log("Reading transaction finished.")
+		case <-w.stopCh:
+			w.t.Logf("%q finished.", w.name())
 			return rs, nil
 		default:
 		}
 
-		err := r.db.View(func(tx *bolt.Tx) error {
-			b := tx.Bucket(r.bucket)
+		op := w.pickOperation()
+		rec, err := runOperation(op, w.db, w.bucket, w.keys, w.conf)
+		if err != nil {
+			readErr := fmt.Errorf("[%s: %s]: %w", w.name(), op, err)
+			w.t.Error(readErr)
+			w.errCh <- readErr
+			return rs, readErr
+		}
 
-			selectedKey := r.keys[mrand.Intn(len(r.keys))]
-			initialVal := b.Get([]byte(selectedKey))
-			time.Sleep(randomDurationInRange(r.readTime.min, r.readTime.max))
-			val := b.Get([]byte(selectedKey))
+		rs = append(rs, rec)
+		time.Sleep(randomDurationInRange(w.conf.workInterval.min, w.conf.workInterval.max))
+	}
+}
 
-			if !reflect.DeepEqual(initialVal, val) {
-				return fmt.Errorf("read different values for the same key (%q), value1: %q, value2: %q",
-					selectedKey, formatBytes(initialVal), formatBytes(val))
-			}
+func (w *worker) pickOperation() OperationType {
+	sum := 0
+	for _, op := range w.conf.operationRatio {
+		sum += op.chance
+	}
+	roll := mrand.Int() % sum
+	for _, op := range w.conf.operationRatio {
+		if roll < op.chance {
+			return op.operation
+		}
+		roll -= op.chance
+	}
+	panic("unexpected")
+}
 
-			clonedVal := make([]byte, len(val))
-			copy(clonedVal, val)
+func runOperation(op OperationType, db *btesting.DB, bucket []byte, keys []string, conf concurrentConfig) (historyRecord, error) {
```

Comment:
As all other operations have matching prefix, like `executeRead` `executeWrite` `executeDelete` I would suggest renaming this function to `executeOperation` as followup.

Reply:
https://github.com/etcd-io/bbolt/pull/471